internal/dataplane/kong_client.go

```diff
@@ -439,6 +452,50 @@ func (c *KongClient) Update(ctx context.Context) error {
 	return nil
 }
 
+func (c *KongClient) FetchLastGoodConfiguration(ctx context.Context) error {
+	if c.lastValidKongState == nil {
+		gatewayClients := c.clientsProvider.GatewayClients()
+		c.logger.Debugf("sending configuration to %d gateway clients", len(gatewayClients))
+
+		type kongStateWithHash struct {
+			kongState *kongstate.KongState
+			hash      string
+		}
+
+		var goodKongState *kongStateWithHash
+		_, err := iter.MapErr(gatewayClients, func(client **adminapi.Client) (*kongStateWithHash, error) {
+			rs, err := dump.Get(ctx, (*client).AdminAPIClient(), dump.Config{})
```

Comment:
Why aren't we using the `go-kong.Client.Config()` added in https://github.com/Kong/go-kong/pull/354 here to retrieve the configuration?

Reply:
Do we really need pointer to pointer?

Reply:
I used it in the first implementation before finding out we have `dump.Get`. `dump.Get` is more structured, and fetches all the needed information. Since this call is performed only at startup, I don't see performance-related problems.

Reply:
Understood. 👍 

One thing that bugs me still is that by using `dump.Get` we have to maintain code that translates `deck.KongRawState` to `KongState`. I can imagine we can easily forget to add code for new entities there without proper testing. This will be hard to catch as it's quite a special case triggered only as a failover and regular E2Es or integration tests won't catch it.

I'm wondering if we could use some `reflect` to ensure we have code translating every entity collection that's in `KongRawState`. In a test, we could iterate over the `KongRawState` fields and ensure that we have at least one testcase that covers it. Something like that:

```go
func TestKongStateToKongRawState_Ensure(t *testing.T) {
	kongRawStateFieldsKICDoesntSupport := []string{
		// These are fields that KIC explicitly doesn't support.
		"Vault",
	}
	allKongRawStateFields := func() []string {
		var fields []string
		typ := reflect.ValueOf(utils.KongRawState{}).Type()
		for i := 0; i < typ.NumField(); i++ {
			fields = append(fields, typ.Field(i).Name)
		}
		return fields
	}()

	testCases := []struct {
		testedFields []string
		// ... input, output, etc. fields ...
	}{
		{
			testedFields: []string{"Services"},
		},
	}

	// Kinda meta test - ensure we have testcases covering all fields in KongRawState.
	for _, field := range allKongRawStateFields {
		if lo.Contains(kongRawStateFieldsKICDoesntSupport, field) {
			t.Logf("skipping field %s - unsupported explicitly", field)
			continue
		}
		testCoveringFieldExists := lo.ContainsBy(testCases, func(tc struct{ testedFields []string }) bool {
			return lo.Contains(tc.testedFields, field)
		})
		assert.True(t, testCoveringFieldExists, "no test covering field %s", field)
	}

	// Run the tests.
	for _, tc := range testCases {
		// ...
		_ = tc
	}
}

```

Reply:
yeah, it makes sense! I've introduced it here: 5fb3a981b7c5fe40bd11ce0bec92f2e5f495e404

Reply:
Thank you :) When looking at it again, I thought that we may even programmatically determine which fields are actually tested in our suite and use the actually tested fields to be compared against all existing ones. What would you say about [that](https://github.com/Kong/kubernetes-ingress-controller/commit/760c30ca322f62fae7039437320311d1ecc425d9)?

Reply:
Sounds good, it looks good to me! Feel free to push it on this branch :)